cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+    try:
+        container_type = ContainerType.from_source_olx_tag(source_node.tag)
+    except ValueError:
+        container_type = None
+        if source_node.tag in {"course", "library"}:
+            should_migrate_node = False
+            should_migrate_children = True
+        else:
+            should_migrate_node = True
+            should_migrate_children = False
+    else:
+        node_level = CompositionLevel(container_type.value)
+        should_migrate_node = not node_level.is_higher_than(context.composition_level)
+        should_migrate_children = True
+    migrated_children: list[_MigratedNode] = []
+    if should_migrate_children:
+        migrated_children = [
+            _migrate_node(
+                context=context,
+                content_by_filename=content_by_filename,
+                source_node=source_node_child,
+            )
+            for source_node_child in source_node.getchildren()
+        ]
+    source_to_target: tuple[UsageKey, PublishableEntityVersion] | None = None
+    if should_migrate_node:
+        source_olx = etree.tostring(source_node).decode('utf-8')
+        if source_block_id := source_node.get('url_name'):
+            source_key: UsageKey = context.source_context_key.make_usage_key(source_node.tag, source_block_id)
+            target_entity_version = (
+                _migrate_container(
+                    context=context,
+                    source_key=source_key,
+                    container_type=container_type,
+                    title=source_node.get('display_name', source_block_id),
+                    children=[
+                        migrated_child.source_to_target[1]
+                        for migrated_child in migrated_children if
+                        migrated_child.source_to_target
+                    ],
+                )
+                if container_type else
+                _migrate_component(
+                    context=context,
+                    content_by_filename=content_by_filename,
+                    source_key=source_key,
+                    olx=source_olx,
+                )
+            )
+            if target_entity_version:
+                source_to_target = (source_key, target_entity_version)
+                # Update the existing_source_to_target_keys dictionary with the new mapping
+                #
+                # This step is critical because it ensures we have an up-to-date mapping of source keys
+                # to target entities.
+                #
+                # While we could re-fetch entity versions from the database to guarantee this,
+                # it's currently unnecessary as the required target entities are actually available.
+                #
+                # @@TODO: Implement a stricter enforcement mechanism.
+                # The system should prevent a node from being migrated without an update
+                # to the `existing_source_to_target_keys` dictionary.
```

Comment:
Seems OK to me if we don't have a stricter enforcement mechanism. If MigrationContext were meant to be used across different edx-platform apps, then it'd be important to be more strict. But since it just is used for tasks.py, I think what you have now is fine.